toolkit/tools/pkg/imagecustomizerlib/customizeboot.go

```diff
@@ -107,143 +93,228 @@ func replaceToken(inputGrubCfgContent string, oldToken string, newToken string)
-	argumentRegex, err := regexp.Compile(argumentRegexString)
+// - args: A list of kernel command-line arguments.
+// - insertToken: A tokenizer token that represents an appropriate insert point for any new args.
+func getLinuxCommandLineArgs(grub2Config string) ([]grubConfigLinuxArg, grub.Token, error) {
+	linuxLine, err := findLinuxLine(grub2Config)
 	if err != nil {
-		return 0, 0, "", fmt.Errorf("failed to compile regular expression to find (%s) kernel arg:\n%w", name, err)
+		return nil, grub.Token{}, err
 	}
 
-	match := argumentRegex.FindStringSubmatchIndex(grub2Config)
-	if match == nil {
-		// Arg not found.
-		return -1, -1, "", nil
-	}
+	// Skip the "linux" command and the kernel binary path arg.
+	argTokens := linuxLine[2:]
 
-	argValue := ""
-	if len(match) > argValueGroupEndIndex {
-		argValueStartIndex := match[argValueGroupStartIndex]
-		argValueEndIndex := match[argValueGroupEndIndex]
-		argValue = grub2Config[argValueStartIndex:argValueEndIndex]
-	}
+	args := []grubConfigLinuxArg(nil)
+	insertAtToken := (*grub.Token)(nil)
 
-	argStartIndex := match[argGroupStartIndex]
-	argEndIndex := match[argGroupEndIndex]
+	for i := range argTokens {
+		argToken := argTokens[i]
+		if argToken.Type != grub.WORD {
+			return nil, grub.Token{}, fmt.Errorf("unexpected token (%s) in grub config linux command",
+				grub.TokenTypeString(argToken.Type))
+		}
 
-	return argStartIndex, argEndIndex, argValue, nil
-}
+		if len(argToken.SubWords) == 1 &&
+			argToken.SubWords[0].Type == grub.VAR_EXPANSION &&
+			argToken.SubWords[0].Value == "kernelopts" {
+			// Found the $kernelopts arg.
+			// Any new args to be inserted, will be inserted immediately before this token.
+			insertAtToken = &argToken
+		}
 
-func replaceKernelCommandLineArgumentValue(inputGrubCfgContent string, name string, value string,
-) (outputGrubCfgContent string, oldValue string, err error) {
-	// Search for the arg.
-	start, end, oldValue, err := findKernelCommandLineArgumentArg(inputGrubCfgContent, name)
-	if err != nil {
-		return "", "", err
+		hasVarExpansion := false
+		argStringBuilder := strings.Builder{}
+
+	subWordsLoop:
+		for _, subword := range argToken.SubWords {
+			switch subword.Type {
+			case grub.KEYWORD_STRING, grub.STRING:
+				argStringBuilder.WriteString(subword.Value)
+
+			case grub.QUOTED_VAR_EXPANSION, grub.VAR_EXPANSION:
+				// There is a variable expansion (e.g. $a) in the arg.
+				// Stop here because we don't know what value to fill in for the variable.
+				hasVarExpansion = true
+				break subWordsLoop
+			}
+		}
+
+		argString := argStringBuilder.String()
+		name, value, foundEqSymbol := strings.Cut(argString, "=")
+		if !foundEqSymbol && hasVarExpansion {
+			// Arg has a variable expansion (e.g. $a) that is part of the arg name.
+			// There isn't any easy way to handle such args. So, just ignore them.
+			continue
+		}
+
+		if hasVarExpansion {
+			// The arg string value isn't known because it contains a variable expansion.
+			value = ""
```

Comment:
why set it to empty string and not just keep the `$variable`?

Reply:
Added clarification to code comment.